### PR TITLE
Added in flag formatting.

### PR DESCRIPTION
**Pull request by PythonCoderAS**
## Summary

This pull request adds in support for adding in flag arguments to the signature string produced by a command.

- I decided to refactor most of the argument-specific logic into a private method so that it was more manageable.
- In order to distinguish a flag argument from a normal argument, the flag's prefix and delimiter are added.
- Flags that are literal arguments need to show their name.
    - I decided to borrow from argparse and show the literal values inside of braces.

## Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/ext/commands/core.py [hunk 1]
```diff
@@ -960,47 +962,95 @@ def signature(self):
...
-                name = '|'.join(f'"{v}"' if isinstance(v, str) else str(v) for v in annotation.__args__)
-            if param.default is not param.empty:
-                # We don't want None or '' to trigger the [name=value] case and instead it should
-                # do [name] since [name=None] or [name=] are not exactly useful for the user.
-                should_print = param.default if isinstance(param.default, str) else param.default is not None
-                if should_print:
-                    result.append(f'[{name}={param.default}]' if not greedy else
-                                  f'[{name}={param.default}]...')
-                    continue
+            flag = isinstance(param.annotation, type) and issubclass(param.annotation, FlagConverter)
```
Comment (Rapptz) [inline discord/ext/commands/core.py:965]: For performance reasons this check is done using `hasattr(x, '__comands_is_flag__')`.

File: discord/ext/commands/core.py [hunk 2]
```diff
@@ -960,47 +962,95 @@ def signature(self):
...
+                    flag_name = f"{prefix}{flag_name}"
+
+                    # The part method wants a parameter instance.
+                    flag_default = inspect.Parameter.empty if flag_instance.default is MISSING else flag_instance.default
+                    flag_annotation = inspect.Parameter.empty if flag_instance.annotation is MISSING else flag_instance.annotation
+
+                    # Note: I chose "POSITIONAL_OR_KEYWORD" because it's the normal kind of argument.
+                    flag_param = inspect.Parameter(original_flag_name, kind=inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                                                   default=flag_default, annotation=flag_annotation)
+                    result.append(self._signature_part(flag_name, flag_param, flag=True, delimiter=delimiter))
```
Comment (Rapptz) [inline discord/ext/commands/core.py:985]:
```suggestion
                    result.append(self.__signature_for_param(flag_name, flag_param, flag=True, delimiter=delimiter))
```

File: discord/ext/commands/core.py [hunk 3]
```diff
@@ -960,47 +962,95 @@ def signature(self):
...
+            else:
+                result.append(self._signature_part(name, param))
```
Comment (Rapptz) [inline discord/ext/commands/core.py:987]:
```suggestion
                result.append(self.__signature_for_param(name, param))
```

File: discord/ext/commands/core.py [hunk 4]
```diff
@@ -960,47 +962,95 @@ def signature(self):
...
+
+        return ' '.join(result)
+
+    def _signature_part(self, name, param, flag=False, delimiter="="):
```
Comment (Rapptz) [inline discord/ext/commands/core.py:991]:
```suggestion
    def __signature_for_param(self, name, param, *, flag=False, delimiter="="):
```

**Review by Rapptz — CHANGES_REQUESTED** [review]
This PR lacks some information such as what it's doing. It vaguely mentions "formatting" yet it ended up being implementing signature for flags. It never showed how the formatting ended up looking so bikeshedding was not easily possible. It was not discussed in the `#bikeshedding` channel either to get some wider opinions on how the signature was going to be implemented. In fact the entire reason it wasn't implemented was because we needed to have a design discussion on how we wanted it to look, which this PR completely bypassed.

**Comment by PythonCoderAS** [comment]
Sorry, I did not know the proper procedures for implementing a new design structure, and I also thought that the flag signature formatting was a minor change. I will bring up the topic on the bikeshedding channel. Wherever possible, I wanted to borrow from the `argparse` strategy of formatting.

**Pull request closed without merging**